diff --git a/faiss/CMakeLists.txt b/faiss/CMakeLists.txt
index fd3ddb3..28592a1 100644
--- a/faiss/CMakeLists.txt
+++ b/faiss/CMakeLists.txt
@@ -248,11 +248,11 @@ if(MKL_FOUND)
   target_link_libraries(faiss PRIVATE ${MKL_LIBRARIES})
   target_link_libraries(faiss_avx2 PRIVATE ${MKL_LIBRARIES})
 else()
-  find_package(BLAS REQUIRED)
+  find_package(BLAS REQUIRED PATHS ${CMAKE_INSTALL_DATAROOTDIR}/lapack)
   target_link_libraries(faiss PRIVATE ${BLAS_LIBRARIES})
   target_link_libraries(faiss_avx2 PRIVATE ${BLAS_LIBRARIES})
 
-  find_package(LAPACK REQUIRED)
+  find_package(LAPACK REQUIRED PATHS ${CMAKE_INSTALL_DATAROOTDIR}/lapack)
   target_link_libraries(faiss PRIVATE ${LAPACK_LIBRARIES})
   target_link_libraries(faiss_avx2 PRIVATE ${LAPACK_LIBRARIES})
 endif()
diff --git a/faiss/impl/HNSW.cpp b/faiss/impl/HNSW.cpp
index 1ed15d7..b356da9 100644
--- a/faiss/impl/HNSW.cpp
+++ b/faiss/impl/HNSW.cpp
@@ -837,8 +831,10 @@ void HNSW::MinimaxHeap::push(storage_idx_t i, float v) {
     if (k == n) {
         if (v >= dis[0])
             return;
+        if (ids[0] != -1) {
+            --nvalid;
+        }
         faiss::heap_pop<HC>(k--, dis.data(), ids.data());
-        --nvalid;
     }
     faiss::heap_push<HC>(++k, dis.data(), ids.data(), v, i);
     ++nvalid;
